text/0000-dockerfiles.md

```diff
@@ -0,0 +1,218 @@
+# Meta
+[meta]: #meta
+- Name: Support Dockerfiles
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks ([RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md)).
+
+# Motivation
+[motivation]: #motivation
+
+[RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md) introduces complexity by defining an API that allows buildpacks to modify base images. To avoid this complexity, we could rely on generated Dockerfiles for base image manipulation. This would simplify the original proposal by, e.g., only requiring a copy of the buildpack on the build-time base image.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with dynamically-generated build-time and runtime Dockerfiles that act as pre-build hooks.
+These hooks would participate in detection and execute before the buildpack build process.
+
+For a given application, a build that uses hooks could be optimized by creating a more narrowly-scoped builder that does not contain hooks.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### Dynamically-applied Dockerfiles
+
+A builder image may include any number of "hook" directories in `/cnb/hooks/`.
+
+Hooks are similar to buildpacks: they have a `/bin/build` and `/bin/detect` executable.
+However, instead of a `buildpack.toml` file, hooks have a `hook.toml` file:
+```toml
+api = "<buildpack API version>"
+
+[hook]
+id = "<hook ID>"
+name = "<hook name>"
+version = "<hook version>"
+homepage = "<hook homepage>"
+description = "<hook description>"
+keywords = [ "<string>" ]
+
+[[hook.licenses]]
+type = "<string>"
+uri = "<uri>"
+```
+
+Hooks may be packaged and examined similar to buildpacks, but with analogous `pack hook` subcommands.
+
+Other pack CLI commands, such as `pack builder create`, will be extended to include support for hooks.
+
+Unlike buildpacks,
+- Hooks must not be included in a meta-buildpacks
+- Hooks must not have `order`/`group` definitions in `hook.toml`
+
+Hooks participate in the buildpack detection process, with the same interface for `/bin/detect`.
+However,
+- `/bin/detect` is optional for hooks, and they are assumed to pass detection when it is not presetn.
+- Hooks may only output `provides` entries to the build plan. They must not output `requires`.
+- Hooks must all proceed regular buildpacks in `order` definitions (e.g., in `builder.toml`).
```

Comment:
This might be my lack of understanding of the English language but what does "proceed" here mean? Does there need to be more clarification?

Reply:
I think this is meant to be `precede`

Reply:
Can we get a `builder.toml` example somewhere in the rfc? Is there a new `[hooks]` table?

Reply:
A separate `[[extensions]]` table seems ideal to me, given that all extensions must come first.

This RFC isn't intended to be exhaustively cover the pack CLI UX, but if you have strong opinions about `builder.toml`, could you propose a suggestion?